v2/snmp/session.go

```diff
@@ -0,0 +1,329 @@
+// the ASN.1 transport format.
+type PDU struct {
+	RequestID int32
+	// Non-zero used to indicate that an exception occurred to prevent the processing of the request
+	Error int
+	// If Error is non-zero, identifies which variable binding in the list caused the exception
+	ErrorIndex  int
+	VarbindList []Varbind
+}
+
+type Varbind struct {
+	OID        asn1.ObjectIdentifier
+	TypedValue *TypedValue
+}
+
+// Encapsulates the data type and value of a variable received in a variable binding from an agent.
+type TypedValue struct {
+	Type  DataType
+	Value interface{}
+}
+
+type sessionImpl struct {
+	conn          net.Conn
+	config        *sessionConfig
+	nextRequestID int32
+}
+
+// rawPDU defines the pdu that is used to passed to/from an SNMP agent.
+type rawPDU struct {
+	RequestID int32
+	// Non-zero used to indicate that an exception occurred to prevent the processing of the request
+	Error int
+	// If Error is non-zero, identifies which variable binding in the list caused the exception
+	ErrorIndex  int
+	VarbindList []rawVarbind
+}
+
+type rawVarbind struct {
+	OID   asn1.ObjectIdentifier
+	Value asn1.RawValue
+}
+
+// Defines the SNMP packet passed over the network to/from an SNMP agent.
+// Note the pdu is initially unmarshalled as a raw value, so that the SNMP message type can be replaced by
+// the ASN1 sequence tag before the variable bindings it contains are unmarshalled.
+type packet struct {
+	Version   SNMPVersion
+	Community []byte
+	RawPdu    asn1.RawValue
+}
+
+const maxInputBufferSize = 65535
+
+// Supported SNMP message types.
+type messageType byte
+
+const getMessage = 0xA0
+const getNextMessage = 0xA1
+const getBulkMessage = 0xA5
+
+func (m *sessionImpl) Get(ctx context.Context, oids []string) (*PDU, error) {
+	return m.executeGet(ctx, getMessage, oids, 0, 0)
+}
+
+func (m *sessionImpl) GetNext(ctx context.Context, oids []string) (*PDU, error) {
+	return m.executeGet(ctx, getNextMessage, oids, 0, 0)
+}
+
+func (m *sessionImpl) GetBulk(ctx context.Context, oids []string, nonRepeaters, maxRepetitions int) (*PDU, error) {
+	return m.executeGet(ctx, getBulkMessage, oids, nonRepeaters, maxRepetitions)
+}
+
+func (m *sessionImpl) Walk(ctx context.Context, rootOid string, walker Walker) error {
+	return m.executeWalk(ctx, getNextMessage, 0, rootOid, walker)
+}
+
+func (m *sessionImpl) BulkWalk(ctx context.Context, rootOid string, maxRepetitions int, walker Walker) error {
+	return m.executeWalk(ctx, getBulkMessage, maxRepetitions, rootOid, walker)
+}
+
+// Generic Get execution.
+// Generates a packet to define the type of Get, the required oids and, in the case of a bulk get, the associated
+// non-repeaters and max-repetitions values.
+// Returns a PDU with the resolved variable bindings.
+func (m *sessionImpl) executeGet(ctx context.Context, getType messageType, oids []string, nonRepeaters, maxRepetitions int) (*PDU, error) {
+
+	// TODO Validate OIDs on entry.
```

Comment:
are you wanting to leave these todos in?

Reply:
Prefer to, for the time being.